conf.d/main

```diff
@@ -1,5 +1,8 @@
 #!/bin/sh -ex
 
+APT_GPG_URL=https://syncthing.net/release-key.txt
+curl -sS $APT_GPG_URL | apt-key --keyring /usr/share/keyrings/syncthing.gpg add -
```

Comment:
Unless apt-key function has changed this is superflous (and potentially dangerous). IIRC apt-key add is to be used in the case that we want to add a new ultimately trusted gpg key for any/all repos, whereas we only want this key to be used for verifying the syncthing repo.

This page doesn't explain it but does warn about not using `apt-key add` on keys for third party repos https://wiki.debian.org/DebianRepository/UseThirdParty

Reply:
My understanding is that using the `--keyring /usr/share/keyrings/syncthing.gpg` writes to the specific keyfile noted, rather than an ultimately trusted keyring (in `/etc/apt/trsuted.gpg` and/or in `/etc/apt/trusted.gpg.d/`), thus addressing your point. AFAIK without the `--keyring` switch it will write to `/etc/apt/trusted.gpg`.

I.e. this command:
```
curl -sS $APT_GPG_URL | apt-key --keyring /usr/share/keyrings/syncthing.gpg add -
```
should essentially be the same as:
```
curl -s $APT_GPG_URL > /usr/share/keyrings/syncthing.gpg
```
Except that when passed through apt-key, the key/gpg/whatever file is de-amoured if need be, or left as is if not needed.

Note that the apt line is:
```
deb [signed-by=/usr/share/keyrings/syncthing.gpg] https://apt.syncthing.net/ syncthing stable
```
Although aving said all that, I should probably test it and report back...

Reply:
FWIW, it looks like my understanding is correct:
```
root@tkldev4 .../products/syncthing# fab-chroot build/root.patched/
root@tkldev4 /# mv /usr/share/keyrings/syncthing.gpg /usr/share/keyrings/syncthing.gpg.bak
root@tkldev4 /# apt update
Get:1 http://deb.debian.org/debian buster InRelease [121 kB]
Get:2 http://security.debian.org buster/updates InRelease [65.4 kB]                                           
Ign:3 http://archive.turnkeylinux.org/debian buster-security InRelease                                                                                    
Get:4 http://security.debian.org buster/updates/main amd64 Packages [208 kB]                
Get:5 http://security.debian.org buster/updates/main Translation-en [113 kB]                                          
Ign:6 http://archive.turnkeylinux.org/debian buster InRelease                                                                         
Get:7 http://deb.debian.org/debian buster/main amd64 Packages [7905 kB]                               
Get:8 http://archive.turnkeylinux.org/debian buster-security Release [3857 B]
Get:9 http://archive.turnkeylinux.org/debian buster Release [3830 B]                   
Get:10 http://archive.turnkeylinux.org/debian buster-security Release.gpg [833 B]     
Get:11 https://apt.syncthing.net syncthing InRelease [15.1 kB]
Get:12 http://archive.turnkeylinux.org/debian buster Release.gpg [833 B]                   
Get:13 http://archive.turnkeylinux.org/debian buster/main amd64 Packages [30.8 kB]                     
Err:11 https://apt.syncthing.net syncthing InRelease                 
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY D26E6ED000654A3E
Get:14 http://deb.debian.org/debian buster/main Translation-en [5969 kB]          
Get:15 http://deb.debian.org/debian buster/contrib amd64 Packages [51.0 kB]                    
Get:16 http://deb.debian.org/debian buster/contrib Translation-en [44.7 kB]
Reading package lists... Done                                                                                                                                                                                     
W: GPG error: https://apt.syncthing.net syncthing InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY D26E6ED000654A3E
E: The repository 'https://apt.syncthing.net syncthing InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
root@tkldev4 /# mv /usr/share/keyrings/syncthing.gpg.bak /usr/share/keyrings/syncthing.gpg
root@tkldev4 /# apt update
Hit:1 http://deb.debian.org/debian buster InRelease
Hit:2 http://security.debian.org buster/updates InRelease                                                         
Ign:3 http://archive.turnkeylinux.org/debian buster-security InRelease                                            
Ign:4 http://archive.turnkeylinux.org/debian buster InRelease            
Hit:5 http://archive.turnkeylinux.org/debian buster-security Release     
Get:7 https://apt.syncthing.net syncthing InRelease [15.1 kB]                   
Hit:8 http://archive.turnkeylinux.org/debian buster Release                     
Get:10 https://apt.syncthing.net syncthing/stable amd64 Packages [1902 B]
Fetched 1902 B in 2s (916 B/s)      
Reading package lists... Done
Building dependency tree       
Reading state information... Done
All packages are up to date.
```

Reply:
Hmm interesting I wasn't aware apt-key had this functionality, that sounds reasonable, however I believe we should first test, with the key present and removing the `[signed-by=...]` parameter. The issue after all is that traditional apt-key usage makes the key ultimately trusted for all repos (iirc). Your evidence above does seem to suggest that using --keyring negates this effect, however without testing with a removed `[signed-by=...]` we can't be sure. Removing the key, may simply be proving that it fails to verify against a key it can't find, it does not necessarily prove the key wont verify packages it shouldn't.

I propose that if you can `apt update` successfully with `apt-key add --keyring ...` and with `[signed-by=...]` removed from the sources list then the issue still persists and we should revert back to the previous method of adding debian keys. If `apt-update` fails then you are correct and this method should be safe for us to use. (Though I do have personal preference on the matter, I see no concrete issues with it if this test shows you to be correct)

Reply:
> [...] I believe we should first test, with the key present and removing the [signed-by=...] parameter. The issue after all is that traditional apt-key usage makes the key ultimately trusted for all repos (iirc). Your evidence above does seem to suggest that using --keyring negates this effect, however without testing with a removed [signed-by=...] we can't be sure. Removing the key, may simply be proving that it fails to verify against a key it can't find, it does not necessarily prove the key wont verify packages it shouldn't.

Ah, yes... good point! You are right, my test doesn't actually prove that the key doesn't exist in one of the trusted keyrings (as you note; because the `[signed-by=...]` is specifying a keyring patch). I'll try again and post back.

Reply:
Ok, the updated test confirms my understanding:
```
root@tkldev4 .../products/syncthing# fab-chroot build/root.patched/

root@tkldev4 /# cat /etc/apt/sources.list.d/syncthing.list
deb https://apt.syncthing.net/ syncthing stable

root@tkldev4 /# apt update
Hit:1 http://deb.debian.org/debian buster InRelease
Hit:2 http://security.debian.org buster/updates InRelease                                                         
Ign:3 http://archive.turnkeylinux.org/debian buster-security InRelease                                            
Ign:4 http://archive.turnkeylinux.org/debian buster InRelease            
Hit:5 http://archive.turnkeylinux.org/debian buster-security Release     
Hit:7 http://archive.turnkeylinux.org/debian buster Release                     
Hit:8 https://apt.syncthing.net syncthing InRelease                             
Err:8 https://apt.syncthing.net syncthing InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY D26E6ED000654A3E
Reading package lists... Done
Building dependency tree       
Reading state information... Done
All packages are up to date.
W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://apt.syncthing.net syncthing InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY D26E6ED000654A3E
W: Failed to fetch https://apt.syncthing.net/dists/syncthing/InRelease  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY D26E6ED000654A3E
W: Some index files failed to download. They have been ignored, or old ones used instead.
```
Then to be 100% sure, I reinstated the '[signed-by=...]` bit:
```
root@tkldev4 /# cat /etc/apt/sources.list.d/syncthing.list 
deb [signed-by=/usr/share/keyrings/syncthing.gpg] https://apt.syncthing.net/ syncthing stable

root@tkldev4 /# apt update
Hit:1 http://deb.debian.org/debian buster InRelease
Hit:2 http://security.debian.org buster/updates InRelease                                                   
Ign:3 http://archive.turnkeylinux.org/debian buster-security InRelease                                      
Ign:4 http://archive.turnkeylinux.org/debian buster InRelease
Hit:5 https://apt.syncthing.net syncthing InRelease
Hit:6 http://archive.turnkeylinux.org/debian buster-security Release
Hit:8 http://archive.turnkeylinux.org/debian buster Release
Reading package lists... Done
Building dependency tree       
Reading state information... Done
All packages are up to date.
```
So I'm confident that we're all good! :smile: